domain/modeldefaults/state/state.go

```diff
@@ -230,10 +237,275 @@ WHERE m.uuid = ?
+		for k, v := range updateAttrs {
+			err := tx.Query(ctx, upsertStmt, cloudDefaultValue{UUID: cld.UUID, Key: k, Value: v}).Run()
+			// The cloud UUID has previously been checked. This allows us to avoid having to use RunAtomic.
+			if database.IsErrConstraintForeignKey(err) {
+				return interrors.Errorf("cloud %q not found%w", cloudUUID, errors.Hide(clouderrors.NotFound))
+			} else if err != nil {
+				return interrors.Errorf("updating cloud default keys %q: %w", cloudUUID, err)
+			}
+		}
+		return nil
+	})
+}
+
+// DeleteCloudDefaults deletes the specified cloud default
+// config values if they exist.
+func (st *State) DeleteCloudDefaults(
+	ctx context.Context,
+	cloudUUID cloud.UUID,
+	removeAttrs []string,
+) error {
+	if len(removeAttrs) == 0 {
+		return nil
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	cld := dbCloud{UUID: cloudUUID.String()}
+	toRemove := attrs(removeAttrs)
+
+	deleteStmt, err := st.Prepare(`
+DELETE FROM cloud_defaults
+WHERE key IN ($attrs[:])
+AND cloud_uuid = $dbCloud.uuid;
+`, toRemove, cld)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return tx.Query(ctx, deleteStmt, toRemove, cld).Run()
+	})
+	if err != nil {
+		return interrors.Errorf("removing cloud default keys for %q: %w", cloudUUID, err)
+	}
+	return nil
+}
+
+// UpdateCloudRegionDefaults is responsible for updating default config values
+// for a cloud region. This function will allow the addition and updating of
+// attributes. If the cloud is not found an error satisfying [clouderrors.NotFound]
+// is returned. If the region is not found, am error satisfying [errors.NotFound]
+// is returned.
+func (st *State) UpdateCloudRegionDefaults(
+	ctx context.Context,
+	cloudUUID cloud.UUID,
+	regionName string,
+	updateAttrs map[string]string,
+) error {
+	if len(updateAttrs) == 0 {
+		return nil
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	cld := dbCloud{UUID: cloudUUID.String()}
+	region := cloudRegion{Name: regionName}
+
+	selectStmt, err := st.Prepare(`
+SELECT &cloudRegion.uuid
+FROM cloud_region
+WHERE cloud_region.cloud_uuid = $dbCloud.uuid
+AND cloud_region.name = $cloudRegion.name;
+`, region, cld)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	upsertStmt, err := st.Prepare(`
+INSERT INTO cloud_region_defaults (region_uuid, key, value)
+VALUES ($cloudRegionDefaultValue.*)
+ON CONFLICT(region_uuid, key) DO UPDATE
+    SET value = excluded.value
+    WHERE region_uuid = excluded.region_uuid
+    AND key = excluded.key;
+`, cloudRegionDefaultValue{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, selectStmt, cld, region).Get(&region)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return interrors.Errorf("cloud %q region %q %w", cloudUUID, regionName, errors.NotFound)
```

Comment:
I guess this is copied over from Cloud, but it would be good to get rid of uses of the generic NotFound error, this could be a clouderrors.RegionNotFound instead

Reply:
There isn't a cloud region not found error. The existing upstream code which receives the error is the api caller and we can't change the error type sent over the wire. There's no point in changing this - introducing a new bespoke error would just need to have it detected and a generic not found sent over the wire anyway.